src/v/archival/adjacent_segment_merger.cc

```diff
@@ -47,79 +47,127 @@ adjacent_segment_merger::adjacent_segment_merger(
 
 ss::future<> adjacent_segment_merger::stop() { return _gate.close(); }
 
-ss::future<> adjacent_segment_merger::run(retry_chain_node& rtc) {
+void adjacent_segment_merger::set_enabled(bool enabled) { _enabled = enabled; }
+
+ss::future<housekeeping_job::run_result>
+adjacent_segment_merger::run(retry_chain_node& rtc, run_quota_t quota) {
     ss::gate::holder h(_gate);
-    vlog(_ctxlog.debug, "Adjacent segment merger run begin");
-    auto scanner = [this](
-                     model::offset local_start_offset,
-                     const cloud_storage::partition_manifest& manifest)
-      -> std::optional<adjacent_segment_run> {
-        adjacent_segment_run run(_archiver.get_ntp());
-        auto [low_watermark, high_watermark] = get_low_high_segment_size(
-          _archiver.get_local_segment_size(),
-          _min_segment_size,
-          _target_segment_size);
-        model::offset so = _last;
-        if (so == model::offset{} && _is_local) {
-            // Local lookup, start from local start offset
-            so = local_start_offset;
-        } else {
-            // Remote lookup, start from start offset in the manifest (or 0)
-            so = _archiver.manifest().get_start_offset().value_or(
-              model::offset{0});
+    run_result result{
+      .status = run_status::skipped,
+      .consumed = run_quota_t(0),
+      .remaining = quota,
+    };
+    for (int i = 0; i < max_reuploads_per_run; i++) {
+        if (!_enabled || _as.abort_requested()) {
+            co_return result;
         }
-        vlog(
-          _ctxlog.debug,
-          "Searching for adjacent segment run, start: {}, is_local: {}, "
-          "local_start_offset: {}",
-          so,
-          _is_local,
-          local_start_offset);
-        for (auto it = manifest.segment_containing(so); it != manifest.end();
-             it++) {
-            if (
-              !_is_local && it->second.committed_offset >= local_start_offset) {
-                // We're looking for the remote segment
+        if (result.remaining <= 0) {
+            co_return result;
+        }
+        vlog(_ctxlog.debug, "Adjacent segment merger run begin");
+        auto scanner = [this](
+                         model::offset local_start_offset,
+                         const cloud_storage::partition_manifest& manifest)
+          -> std::optional<adjacent_segment_run> {
+            adjacent_segment_run run(_archiver.get_ntp());
+            auto [low_watermark, high_watermark] = get_low_high_segment_size(
```

Comment:
nit: I think it was feedback from the previous PR, but "watermark" tends to refer to offset instead of size. Maybe rename these to `size_lower_bound` and `size_upper_bound` or somesuch?